QuoridorEngine/src/ui/gui/Board.cs

```diff
@@ -0,0 +1,182 @@
+        public void RemoveEventHandlers()
+        {
+            for (int row = boardCells.GetLength(0) - 1; row >= 0; row--)
+                for (int column = 0; column < boardCells.GetLength(1); column++)
+                    if (boardCells[row, column] is PlayerCell)
+                        ((PlayerCell)boardCells[row, column]).RemoveEventHandlers(eventHandlers["OnPlayerCellClick"]);
+                    else if (boardCells[row, column] is WallPartCell)
+                        ((WallPartCell)boardCells[row, column]).RemoveEventHandlers(
+                            eventHandlers["OnWallCellEnter"], eventHandlers["OnWallCellLeave"], eventHandlers["OnWallCellClick"]);
+
+            whitePawn.RemoveEventHandlers(eventHandlers["OnPlayerPawnClick"]);
+            blackPawn.RemoveEventHandlers(eventHandlers["OnPlayerPawnClick"]);
+        }
+
+        // Coordinate board cells drawing on board panel
+        private void drawBoard()
+        {
+            boardCells = new BoardCell[dimension, dimension];
+            int xLoc = 0, yLoc = 0;
+
+            for (int row = boardCells.GetLength(0) - 1; row >= 0; row--)
+            {
+                for (int column = 0; column < boardCells.GetLength(1); column++)
+                {
+                    drawBoardCell(row, column, xLoc, yLoc);
+                    xLoc += boardCells[row, column].Width;
+                }
+                yLoc += boardCells[row, 0].Height;
+                xLoc = 0;
+            }
+        }
+
+        /// <summary>
+        /// Generate and draw board cells based on their type
+        /// </summary>
+        /// <param name="row">Board row</param>
+        /// <param name="column">Board column</param>
+        /// <param name="xLoc">Cell x location</param>
+        /// <param name="yLoc">Cell y location</param>
+        private void drawBoardCell(int row, int column, int xLoc, int yLoc)
+        {
+            if (row % 2 == 0 && column % 2 == 0)
+                boardCells[row, column] = new PlayerCell(row, column, maxCellSize);
+            else if (row % 2 == 1 && column % 2 == 1)
+                boardCells[row, column] = new WallCornerCell(row, column, minCellSize);
+            else
+                boardCells[row, column] = new WallPartCell(row, column, minCellSize, maxCellSize);
+
+            boardCells[row, column].Location = new Point(xLoc, yLoc);
+            Controls.Add(boardCells[row, column]);
+        }
+
+        /// <summary>
+        /// Render player pawn to board 
+        /// </summary>
+        /// <param name="isWhitePlayer">Player option</param>
+        public void drawPlayerPawn(bool isWhitePlayer)
+        {
+            ref PlayerPawn playerPawn = ref getPlayerPawn(isWhitePlayer);
+            playerPawn = new PlayerPawn(isWhitePlayer, maxCellSize);
+            Controls.Add(playerPawn);
+            playerPawn.BringToFront();
+        }
+
+        private ref PlayerPawn getPlayerPawn(bool isWhitePlayer)
```

Comment:
Classes are passed "by reference" (pointer to class) by default. Why would you need to return a reference to the pointer of the class with this method instead of simply returning the class (pointer to class behind the scenes)?